src/python/pants/backend/go/util_rules/cgo.py

```diff
@@ -0,0 +1,858 @@
+        output_files=(obj_file,),
+        description=description,
+    )
+
+
+@rule_helper
+async def _gccld(
+    binary_name: str,
+    input_digest: Digest,
+    obj_dir_path: str,
+    outfile: str,
+    flags: Iterable[str],
+    objs: Iterable[str],
+    description: str,
+    golang_subsystem: GolangSubsystem,
+) -> FallibleProcessResult:
+    compiler_path = await Get(
+        BinaryPath,
+        CGoBinaryPathRequest(
+            binary_name=binary_name,
+            binary_path_test=BinaryPathTest(["--version"]),
+        ),
+    )
+
+    # TODO: Select CXX if any C++ code is present.
+    compiler_args_result, env = await MultiGet(
+        Get(SetupCompilerCmdResult, SetupCompilerCmdRequest((compiler_path.path,), obj_dir_path)),
+        Get(Environment, EnvironmentRequest(["PATH"])),
+    )
+
+    args = [*compiler_args_result.args, "-o", outfile, *objs, *flags]
+    result = await Get(
+        FallibleProcessResult,
+        Process(
+            argv=args,
+            env={"TERM": "dumb", **env},
+            input_digest=input_digest,
+            output_files=(outfile,),
+            description=description,
+        ),
+    )
+
+    # TODO: Filter out output with irrelevant warnings just like `go` tool does.
+
+    return result
+
+
+@dataclass(frozen=True)
+class _DynImportResult:
+    digest: Digest
+    dyn_out_go: str | None  # if not empty, is a new Go file to build as part of the package.
+    dyn_out_obj: str | None  # if not empty, is a new file to add to the generated archive.
+    use_external_link: bool
+
+
+# dynimport creates a Go source file named importGo containing
+# //go:cgo_import_dynamic directives for each symbol or library
+# dynamically imported by the object files outObj.
+# dynOutObj, if not empty, is a new file to add to the generated archive.'
+#
+# see https://github.com/golang/go/blob/f28fa952b5f81a63afd96c9c58dceb99cc7d1dbf/src/cmd/go/internal/work/exec.go#L3020
+@rule_helper
+async def _dynimport(
+    input_digest: Digest,
+    obj_files: Iterable[str],
+    obj_dir_path: str,
+    cflags: Iterable[str],
+    ldflags: Iterable[str],
+    pkg_name: str,
+    goroot: GoRoot,
+    import_go_path: str,
+    golang_subsystem: GolangSubsystem,
+) -> _DynImportResult:
+    cgo_main_compile_process = await _cc(
+        binary_name=golang_subsystem.cgo_gcc_binary_name,
+        input_digest=input_digest,
+        work_dir=obj_dir_path,
+        src_file=os.path.join(obj_dir_path, "_cgo_main.c"),
+        flags=cflags,
+        obj_file=os.path.join(obj_dir_path, "_cgo_main.o"),
+        description="Compile _cgo_main.c",
+        golang_subsystem=golang_subsystem,
+    )
+    cgo_main_compile_result = await Get(ProcessResult, Process, cgo_main_compile_process)
+    obj_digest = await Get(
+        Digest, MergeDigests([input_digest, cgo_main_compile_result.output_digest])
+    )
+
+    # TODO: Gather .syso files from this package and all (transitive) dependencies.
```

Comment:
Worth a ticket? I'm not sure why this is necessary

Reply:
Yes. `go` specifies that `syso` files will be linked as-is with the other C code. So it is a mandatory part of the spec.

Reply:
https://github.com/pantsbuild/pants/issues/16827